modules/order/src/Entity/OrderItem.php

```diff
@@ -108,6 +110,26 @@ public function getQuantity() {
   /**
    * {@inheritdoc}
    */
+  public function getItemsQuantity() {
```

Comment:
I agree that for now TypedData system returns proper quantity field values just because they are saved "As Is" in the database, i.e.- in decimal type with precision 10 and scale 2. That means even if you enter, let's say "5" integer value on the "Add to cart" quantity field and save it and look into the commerce_order_item.quantity in the database, you'll see this: "5.00". Not an issue so far because the value is rounded or cast to integer value when rendering Shopping Cart Block and Shopping Cart Summary view.

See: modules/cart/src/Plugin/views/field/EditQuantity.php +98
See: modules/cart/src/Plugin/Block/CartBlock.php + 132

OK, but what if the value were entered as a decimal, like this: "5.5"? Nothing bad really if you calculate an order item total at the price $1.99 for 1 (ONE!) item (5.5 * 1.99 = $10.945), but the quantity displayed in the Shopping Cart Block and Shopping Cart Summary view will be "5", not "5.5".

The question is (not mine, the customer's question): Why I see the right total ($10.945) but wrong quantity (5) in my Shopping Cart? Is it a professional site? Or may be it is a fraud one? Why I can't enter decimal value on a Shopping Cart Summary "EditQuantity" field despite on the "Add to Cart" it was allowed?

The main question is: if we allow decimal values on the quantity what the "quantity" is? Is it a (combined) number of items on a particular order item type or it is the amount of something?

The answer is that: they are both. If the value entered is of a decimal type then the number of items on that order item is always === 1, otherwise the number of items is the actual integer value of a quantity field.

Example 1: you obtain only one (not divisible) item if you've bought fiber cable at the quantity === length === 2.5 (any dimension units).
  
Example 2: you obtain 2 (separate) items if you've bought T-shirts at the quantity === items === 2 (items).

OK, but how we decide what the type of number were entered, given that the value in any case saved in the database as decimal?

The cool HTML5 number type input element gives us a solution, specifically the #step attribute. Actually it restricts the value allowed to enter not only per "step" level but per number type level too, with a maximum scale checked.

Example: given step = 1.05, min = 1 and max = 9999 you restricted to the range 1-9997.05 with the first possible value after minimum is 2.05. Look, not 2 or 2.5 or 2.01 or 2.005, etc., but strictly 2.05 with the next value 3.1 and next 4.15 etc.. Our Drupal js checks the value on the client and even does not allow to save a cart displaying pretty pop-ups for a user with an error description and closest valid values that could be entered.

Client side? But, what if ...

It is not an issue because this patch does not rely on a user input taking into account settings on an order item type's quantity made by administrator. Yes, the malicious input can cause weird results on a price totals and number of items in the cart but not broke the whole system except the Shopping Cart of that clever user. For now these "experiments" still could be done because price or tax calculation code taking whatever the quantity number is multiplyed by a raw price amount. I think that it might be useful if this code also call $order_item->getQuantityWidgetSettings() first and checked validity of a quantity in respect to #min, #max and #step settings in a try{} block and throw an Exception if something is wrong. That will give us an integrity for the quantity <-> price_total bundle and for the whole commerce workflow.

See patched: modules/order/src/Entity/OrderItem.php +115